text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,651 @@
+with the package itself (and not about it's listing in Construct Hub). The
+routing is based on metadata configured on the package itself, through the bug
+tracker URL, repository URL, and homepage URL.
+
+## Technical Design
+
+### Overview
+
+![Architecture Diagram](../images/cdk-construct-hub/architecture-diagram.png)
+
+### What is the high level implementation plan?
+
+#### Overview
+
+The Construct Hub is a relatively simple static web application. While the
+dataset it exposes (all Construct Libraries) may be relatively fast moving,
+there is no necessity for newly published packages (or versions thereof) to be
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. **Package Discovery:** The back-end for Construct Hub starts with a custom
+   built event source, which implements the logic necessary to detect new
+   packages that are relevant to the Construct Hub, and sends messages to an SQS
+   queue for further processing. Notifications have the following attributes:
+
+   Name          | Description
+   --------------|--------------------------------------------------------------
+   `assemblyUri` | The URL to an S3 Object containing the `.jsii` assembly included in the package
+   `time`        | The timestamp at which the version was created
+   `integrity`   | An integirity check for the complete record
+
+   The Construct Hub will have permissions to read from (and, if applicable, to
+   decrypt) the S3 bucket that is used to hold the object denoted by
+   `assemblyUri`. That bucket may have a Lifecycle configuration that cleans up
+   objects after a few days, as objects need not be retained there after they
+   have been successfully ingested by the Construct Hub (which will own and
+   maintain a copy of this artifact).
+
+   Construct Hub comes with a built-in integration function that listents to
+   changes on the [npmjs.com registry](https://skimdb.npmjs.com/registry)
+   database. Operators of private instances are able to interface with other
+   package registries (including private ones), such as CodeArtifact
+   repositories, etc.
+
+1. **Ingestion:** A Lambda function then picks up messages from the SQS queue
+   and prepares the artifacts consumed by the front-end application, stored in a
+   dedicated S3 bucket using the following key format:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly.json`
```

Comment:
The **Discovery** function is an integration point - it may be implemented external to the hub. If we're integrating with a 3P artefact repository (CodeArtifact, Artifactory, GH Packages, etc...), authentication may be needed to download the tarball. If the download was to happen in the **Ingestion** function (which is strictly internal to the hub), then we'd need to allow configuring authentication somehow... and I'd rather not.

The alternative is that the Discovery function creates a "hub-readable" copy of the tarball somewhere (e.g: S3) if the package isn't public... But then why are we duplicating that effort in the system?

Reply:
I understand your point about authentication but my concern is about resiliency and timeouts of the ingestion function. A single lambda invocation won't always be able to handle an entire batch of updates, so we will need some form of queuing or recovery or re-execution based on which items have been processed (downloaded, + jsii extracted + queued).

How do you see this addressed? I guess we can split the ingestion into two stages separated by a queue...

Reply:
>an entire batch of updates

The batch size is 1 for this function. It triggers off of an SQS queue (so queuing is built right in there).